synapse/replication/tcp/handler.py

```diff
@@ -352,24 +352,33 @@ def start_replication(self, hs: "HomeServer") -> None:
 
         reactor = hs.get_reactor()
         redis_config = hs.config.redis
-        if hs.config.redis.redis_use_tls:
-            ssl_context_factory = ClientContextFactory(hs.config.redis)
-            reactor.connectSSL(
-                redis_config.redis_host,
-                redis_config.redis_port,
+        if redis_config.redis_path is not None:
```

Comment:
This seems annoying that we have this logic in two places... I guess we end up with two connections, but still that seems kind of weird.

I do wonder why we don't use the lazy versions here though?

Reply:
I think(and I only vaguely understand some of the process), the Pub/Sub system requires two connections because one is for the `REPLICATE` system not allowing anything to be sent on it and the other for sending commands? Here's my basis for this thinking:
https://github.com/matrix-org/synapse/blob/8839b6c2f8b07d5d122a15e79b1ebdbdd5f3e26b/synapse/replication/tcp/handler.py#L339-L341

As far as lazy versions, I'm assuming(again, based on vague understanding) that it needs the connection for the replication to be *right now* and not *whenever-it-gets-around-to-it*. That being said, they both(lazy and this) actually seem to be the same(I'm only slightly sure they end up being both Deferreds). It might be an academic argument, but I'd rather not touch something I don't understand(and would be another PR anyways). I'll play around with it later.

Reply:
(It may have something to do with using the `RedisDirectTcpReplicationClientFactory`? That seems to be ultimately the only difference. The other(lazy version) builds in the `SynapseRedisFactory`. These could probably be refactored in some way to pass in the factory, but then it would just be the `connect[TCP|SSL|UNIX]` directly. So probably fine the way it is.)

Reply:
I don't think we need to make a change in this PR, just kind of a shame to duplicate the logic.